### ISSUE TITLE
Issue with "Element.scrollHeight": …

### ISSUE THREAD

**Issue by EdNutting** [mdn/content#1156]
MDN URL: https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollHeight

#### What information was incorrect, unhelpful, or incomplete?

The test for reaching the scroll limit shown on this page is inaccurate when using the now fairly typical "load from the bottom" scroller technique (often seen in "chat message scrollers"). In Chrome and Firefox, the `scrollTop` value reports a negative number when using an overflowing Flex element with `column-reverse` direction. This results in the test always failing even when going from the initial (aka bottom) scroll position to the opposite extreme (aka top).

The corrected equation ought to be (I believe):

```js
element.scrollHeight - Math.abs(element.scrollTop) === element.clientHeight
```

#### Specific section or headline?

* `Determine if an element has been totally scrolled`
* `scrollHeight Demo` : `JavaScript` : `checkReading` function

#### What did you expect to see?

An equation that detected reaching the opposing limit of scroll to an element's initial scroll position, regardless of the method used to generate the scrollable area.

#### Did you test this? If so, how?

<details>

<summary>
Tested in the latest versions of Chrome and Firefox on Windows using the following React / Chakra-UI code. (Reproduced here freely under CC-0 license).
</summary>

##### LazyLoadingScroller.tsx

```ts
import { Box, Center, Flex, Heading, Spinner } from "@chakra-ui/react";
import React, { useCallback, useEffect, useMemo, useReducer } from "react";

interface LazyLoadingScrollerProps<T> {
    batchSize: number;

    load: (index: number | null, count: number) => Promise<{ nextIndex: number; newItems: Map<number, T> }>;

    isEqual: (x: T, y: T) => boolean;
    renderItem: (item: T) => React.ReactNode;
}

interface ReducerState<T> {
    stateName: "loading" | "idle" | "end-of-feed";
    loadOnIdle: boolean;
    nextIndex: number | null;
    items: Map<number, T> | null;
    lastError: string | null;
    renderedItems: Map<number, { sourceItem: T; node: React.ReactNode }>;
}

type ReducerActions<T> =
    | { name: "start-load" }
    | {
          name: "render-items";
          newItems: Map<number, T>;
          isEqual: (x: T, y: T) => boolean;
          renderItem: (item: T) => React.ReactNode;
          nextIndex: number;
      }
    | { name: "error"; nextError: string };

function reduce<T>(oldState: ReducerState<T>, action: ReducerActions<T>) {
    const newState = { ...oldState, items: oldState.items ?? new Map() };

    if (action.name === "error") {
        newState.lastError = action.nextError;
    }

    if (newState.stateName === "idle" && action.name === "start-load") {
        newState.stateName = "loading";
        newState.lastError = null;
    } else if (newState.stateName === "loading" && action.name === "render-items") {
        if (action.newItems.size > 0) {
            newState.stateName = "idle";
            newState.lastError = null;
            newState.nextIndex = action.nextIndex;
            newState.renderedItems = new Map(newState.renderedItems ?? []);
            action.newItems.forEach((item, key) => {
                newState.items.set(key, item);

                let shouldRender = true;
                const existingItem = newState.renderedItems.get(key);
                if (existingItem) {
                    shouldRender = !action.isEqual(existingItem.sourceItem, item);
                }

                if (shouldRender) {
                    newState.renderedItems.set(key, {
                        sourceItem: item,
                        node: action.renderItem(item),
                    });
                }
            });
        } else {
            newState.stateName = "end-of-feed";
            newState.lastError = null;
            newState.nextIndex = action.nextIndex;
        }
    } else if (newState.stateName !== "idle" && action.name === "start-load") {
        newState.loadOnIdle = true;
    }

    if (newState.stateName === "idle" && newState.loadOnIdle) {
        newState.loadOnIdle = false;
        newState.stateName = "loading";
    }

    return newState;
}

export default function LazyLoadingScroller<T>(props: LazyLoadingScrollerProps<T>): JSX.Element {
    const [state, act] = useReducer<React.Reducer<ReducerState<T>, ReducerActions<T>>>(reduce, {
        stateName: "loading",
        loadOnIdle: false,
        nextIndex: null,
        items: null,
        lastError: null,
        renderedItems: new Map(),
    });
    const innerContainerRef = React.createRef<HTMLDivElement>();

    const scrollInfo = useCallback(() => {
        const c = innerContainerRef.current;
        if (c) {
            return {
                clientHeight: c.clientHeight,
                scrollHeight: c.scrollHeight,
                scrollTop: c.scrollTop,
            };
        }
        return null;
    }, [innerContainerRef]);

    useEffect(() => {
        if (state.stateName === "loading") {
            (async () => {
                try {
                    const next = await props.load(state.nextIndex, props.batchSize);
                    const newItems = next.newItems;
                    const nextIndex = next.nextIndex;

                    act({
                        name: "render-items",
                        isEqual: props.isEqual,
                        newItems,
                        nextIndex,
                        renderItem: props.renderItem,
                    });
                } catch (e) {
                    act({
                        name: "error",
                        nextError: e.message ?? e.toString(),
                    });
                }
            })();
        }
    }, [props, state.nextIndex, state.stateName]);

    const elements = useMemo(() => {
        return [...state.renderedItems.keys()]
            .sort((x, y) => x - y)
            .map((key) => {
                const renderedItem = state.renderedItems.get(key);
                if (renderedItem) {
                    return renderedItem.node;
                }
            });
    }, [state.renderedItems]);

    // TODO: Display error

    useEffect(() => {
        if (state.stateName === "idle") {
            const sc = scrollInfo();
            if (sc) {
                const offset = sc.scrollHeight - Math.abs(sc.scrollTop);
                if (offset === sc.clientHeight) {
                    act({
                        name: "start-load",
                    });
                }
            }
        }
    }, [scrollInfo, state.stateName]);

    return state.items === null ? (
        <Center h="100%">
            <Box>
                <Spinner aria-label="Loading messages" />
            </Box>
        </Center>
    ) : (
        <Flex
            w="100%"
            h="100%"
            overflowX="hidden"
            overflowY="auto"
            flexDir="column"
            justifyContent="flex-end"
            pos="relative"
        >
            <Flex
                ref={innerContainerRef}
                w="100%"
                h="auto"
                overflowX="hidden"
                overflowY="scroll"
                flexDir="column-reverse"
                minH="100%"
                onScroll={() => {
                    if (state.stateName === "idle") {
                        const sc = scrollInfo();
                        if (sc) {
                            const offset = sc.scrollHeight - Math.abs(sc.scrollTop);
                            if (offset === sc.clientHeight) {
                                act({
                                    name: "start-load",
                                });
                            }
                        }
                    }
                }}
            >
                {elements}
                {state.stateName === "loading" ? (
                    <Center py={5} mb={1}>
                        <Spinner message="Loading messages" />
                    </Center>
                ) : undefined}
                {state.stateName === "end-of-feed" ? (
                    <Heading
                        py={5}
                        mb={1}
                        as="h4"
                        fontSize="0.8em"
                        fontStyle="italic"
                        borderBottomWidth={1}
                        borderBottomStyle="solid"
                        borderBottomColor="gray.400"
                    >
                        (No more messages)
                    </Heading>
                ) : undefined}
            </Flex>
            <Box pos="absolute" bottom="0">
                {state.stateName}
            </Box>
        </Flex>
    );
}
```

##### Demo.tsx

```ts
<LazyLoadingScroller<MessageDescriptor>
      batchSize={10}
      load={async (index, count) => {
          if (index === null) {
              index = 0;
          }

          if (index >= 100) {
              return {
                  newItems: new Map(),
                  nextIndex: index,
              }; // End of feed
          }

          await new Promise(resolve => setTimeout(resolve, 500));

          const results = new Map<number, MessageDescriptor>();
          for (let idx = index; idx < index + count; idx++) {
              results.set(idx, {
                  id: idx,
                  message: idx.toLocaleString(),
              });
          }
          return {
              newItems: results,
              nextIndex: index + count + 1,
          };
      }}
      isEqual={(x, y) => x.message === y.message}
      renderItem={(item) => <Box key={"message-" + item.id}>{item.message}</Box>}
  />
```

</details>

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/web/api/element/scrollheight`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollHeight
* GitHub URL: https://github.com/mdn/content/blob/master/files/en-us/web/api/element/scrollheight/index.html
* Report started: 2021-01-10T17:32:28.631Z

</details>

**Comment by Kingdutch** [comment]
I ran into this same issue and can confirm that the Math.abs fixes the problem.

Additionally I noticed that in Safari on OS X (and probably also for mobile browsers) there is no hard boundary for the scrollHeight. This means that `Math.abs(element.scrollTop)` will grow larger than `element.scrollHeight` which causes the left hand side to dip below `element.clientHeight`. Changing the comparison operator to `<=` fixes that.
